docker/diff/diff.py

```diff
@@ -0,0 +1,197 @@
+
+    job_config = bigquery.QueryJobConfig(destination=table_id)
+
+    query_job = client.query(query, job_config=job_config)
+    print("The query submitted is:")
+    print(query)
+    print(f"The destination table is: {table_id}")
+    # return the destination table name
+    try:
+        query_job.result()
+    except GoogleCloudError as err:
+        print(f"There was a {type(err)}")
+        print(err)
+    # return the destination table name
+    return dest_table_name
+
+
+def get_created_rows():
+    cols = get_non_pk_cols(table_name, repo_data_project, repo_dataset_name, repo_client)
+    # use defaults otherwise
+    get_rows(DiffType.CREATE, cols=cols)
+
+
+def get_deleted_rows():
+    cols = get_non_pk_cols(table_name, repo_data_project, repo_dataset_name, repo_client)
+    get_rows(DiffType.DELETE, cols=cols)
+
+
+def get_updated_rows():
+    cols = get_non_pk_cols(table_name, repo_data_project, repo_dataset_name, repo_client)
+    get_rows(DiffType.UPDATE, cols=cols)
+
+
+def get_rows(diff_type: DiffType, cols: Dict[str, str], table_name: str = table_name,
+             dest_data_project: str = dest_data_project, dest_dataset_name: str = dest_dataset_name,
+             repo_data_project: str = repo_data_project, repo_dataset_name: str = repo_dataset_name,
+             release_date: str = release_date, prev_date: str = prev_date, version_col: str = version_col,
+             client: bigquery.Client = repo_client) -> str:
+    # should use the jade client
+    if diff_type is DiffType.CREATE:
+        select_table = "A"
+        join = "LEFT JOIN"
+        on = "ON A.id = B.id"
+        where = "WHERE B.id IS NULL"
+    elif diff_type is DiffType.DELETE:
+        select_table = "B"
+        join = "RIGHT JOIN"
+        on = "ON A.id = B.id"
+        where = "WHERE A.id IS NULL"
+    elif diff_type is DiffType.UPDATE:
+        select_table = "A"
+        join = "JOIN"
+        # for update, we need everything where the ids are the same but ANYTHING at all in the other columns is
+        # different. We join on the id, AND we join on inequality between all other columns, so all the other cols
+        # are OR-ed together. Some of them are ARRAY cols, so we need to convert them to string before comparing.
+        diff_join = " OR ".join([
+            f"ARRAY_TO_STRING(A.{col_name}, \" \") != ARRAY_TO_STRING(B.{col_name}, \" \")"
+            if "ARRAY" in col_type else f"A.{col_name} != B.{col_name}"
+            for col_name, col_type in cols.items()])
+        on = f"ON A.id = B.id AND ({diff_join})"
+        where = ""
+    else:
+        raise ValueError("Invalid DiffType for get_rows")
+
+    selected_cols = ", ".join([f"{select_table}.{col}" for col in tuple(cols)])
+    query = f"""
+    SELECT {select_table}.id, {selected_cols} FROM
+    (SELECT * FROM `{repo_data_project}.{repo_dataset_name}.{table_name}` WHERE {version_col} = "{release_date}") AS A
```

Comment:
I'm not sure about the back tick vs double quotes here... do these quotes have different meanings in BQ? In SQL you usually see single quote `'` for all SQL string literals.

Reply:
They do; the back ticks are specifically for writing out the whole project-name.dataset-name.table-name thing I believe, it's a special difference in their "standard SQL"